client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java

```diff
@@ -88,16 +90,76 @@ static Request bulk(BulkRequest bulkRequest) throws IOException {
         parameters.withTimeout(bulkRequest.timeout());
         parameters.withRefreshPolicy(bulkRequest.getRefreshPolicy());
 
-        // Bulk API only supports newline delimited JSON
-        XContentType xContentType = XContentType.JSON;
-        byte separator = xContentType.xContent().streamSeparator();
-        ContentType requestContentType = ContentType.create("application/x-ndjson");
+        // Bulk API only supports newline delimited JSON or Smile. Before executing
+        // the bulk, we need to check that all requests have the same content-type
+        // and this content-type is supported by the Bulk API.
+        List<XContentType> allowedContentTypes = Arrays.asList(XContentType.JSON, XContentType.SMILE);
+
+        XContentType bulkContentType = null;
+        for (int i = 0; i < bulkRequest.numberOfActions(); i++) {
+            DocWriteRequest<?> request = bulkRequest.requests().get(i);
+            XContentType requestContentType = null;
+
+            boolean supported = true;
+            boolean match = true;
+
+            DocWriteRequest.OpType opType = request.opType();
+            if (opType == DocWriteRequest.OpType.INDEX || opType == DocWriteRequest.OpType.CREATE) {
+                IndexRequest indexRequest = (IndexRequest) request;
+                requestContentType = indexRequest.getContentType();
+
+                supported = allowedContentTypes.contains(requestContentType);
+                if (bulkContentType == null) {
+                    bulkContentType = requestContentType;
+                } else {
+                    match = (requestContentType == bulkContentType);
```

Comment:
I wish we could throw exception here already rather than setting a flag. maybe add a static method that throws the exception so we don't worry about copying the message around?